---
title: "SK-5A4F_Mouse_DESEQ"
author: "Madison Carpenter"
date: "2023-06-28"
output: html_document
---

# 1.0 Setup the Environment

```{r Setup the Environment, echo=FALSE, message=FALSE}
suppressPackageStartupMessages({
# Load packages
library(tidyverse)
library(magrittr)
library(org.Mm.eg.db)
library(factoextra)
library(gt)
library(clusterProfiler)
})

# Setup Pathways
workdir <- "~/Dropbox (Partners HealthCare)/"
scriptDir <- "Scripts/BulkRNAseq/"

# Set version control
set.seed(123)

# # Load DESeq workflow scripts
source(file = paste0(workdir,scriptDir,"DESeq.R"))
source(file = paste0(workdir,scriptDir,"DataVisualizations.R"))
source(file = paste0(workdir,scriptDir,"Deseq_Statistics.R"))
source(file = paste0(workdir,scriptDir, "pathway_graphs.R"))
source(file = paste0(workdir,scriptDir, "GoEnrichAnalysis.R"))

# Set Diagram settings
options(width=200)
```

# 2.0 Figure 1B and Supplementary Figure 3
## 2.1 Load Data: Xenon vs. Air
```{r pressure, echo=FALSE}
#_____________UPDATE!______________#
# Setup Variables
expdir <- "SK-5A4F_SN0268732_Wesley_Nov_2022_Xenon_Human_Mouse/"

# Load Data
meta <- readxl::read_excel(paste0(workdir, expdir,"SK-5A4F_metadata.xlsx"))
quantfiles <- file.path(workdir,expdir,"transcripts_quant/", meta$Position, "quant.sf")
#_________________________________#
```

## 2.2 DESEQ
### 2.2.1 Modify Metadata 
```{r, echo=FALSE}
##________UPDATE ENTIRE SECTION___________##
# Set a file prefix
fileprefix <- "SK-5A4F_APPPS1_Mouse_XenvsAir" # Experimental group should be listed first in the file prefix**

# Extract Samples Meta Data & Filter Outliers
sub_meta <- meta %>% filter(Genotype == "APPPS1")

# Pull specific Quant files
position_list <- paste0(sub_meta$Position, collapse = '|')
sample_files <- grep(position_list, quantfiles, value = TRUE) # Check number of files matches # of samples

# Check the metadata and files are in the correct order
all(file.exists(quantfiles))

## Create Graphs All 
graph_lab <- paste0(sub_meta$Position,"\n",sub_meta$Sex, "_", sub_meta$Genotype,"\n",sub_meta$CellType,"\n",sub_meta$Treatment)
table_lab <- paste0(sub_meta$Position,"_",sub_meta$Sex, "_", sub_meta$Genotype,"_",sub_meta$CellType,"_",sub_meta$Treatment)


# Format the Meta Data
table_header <- "APPPS1 Mouse: Xenon vs Air"
sub_meta %>%
  gt() %>% 
  tab_header(
    title = md("**Meta Data for Human SK-5A4F samples**"),
    subtitle = md(paste0(table_header, " experiment's meta data are presented"))
  ) %>%
  cols_move_to_start(
    columns = c(Position, Treatment) # Update based on columns **
  )
```

### 2.2.2 Tximport/DESeq

```{r, echo=FALSE}
#_____________UPDATE!______________#
# Setup the Design for Tximport/DESeq2
design = "Treatment"
treatcolumn = "Treatment"
control = "AIR"
refanimal = "mouse"
type <- "LRT"
#_________________________________#

# Run Tx2gene
tx2gene <- tx2genefun(reference = refanimal)

# Run Tximport and DESeq 
ddsTxi1 <- Tximportfun(sample_files,reference = refanimal, metadata = sub_meta, design = design, tx2gene = tx2gene)
dds1 <- DESeqFun(treatcolumn = treatcolumn, control = control, DESeqObject = ddsTxi1, test_type = type)

# Save Deseq output
ddsobj <- dds1[[1]] # DESeq Object
ddsres <- dds1[[2]]# Results
```

### 2.2.3 Export DESEQ results

```{r}
#_____________UPDATE!______________#
resdir <- paste0(workdir,expdir,"/results/Madison/APPPS1/")
#__________________________________#

# Extract the stats on the DESeq run 
extractedstats <-  extractstats(ddsobj, ContrastObj = ddsres ,sub_meta, graphlabel = table_lab)

# Save to variables
DEResults <- extractedstats[[1]] 
DEcounts <- extractedstats[[2]] 
sigDEResults <- extractedstats[[3]]
sigDEcounts <- extractedstats[[4]] 

summary(ddsres, alpha = 0.05)

# Export the Stats
exportstats(DEResults = DEResults, DEcounts = DEcounts,sigDEResults = sigDEResults, 
            sigDEcounts = sigDEcounts, directory = resdir,fileprefix = fileprefix)
```


## 2.3 Data Exploration
### 2.3.1 ScatterPlot Volcano (Figure 1B)

```{r}
# Setup Volcano Scatter Theme
theme_scatter <- theme(aspect.ratio = 1, 
                       panel.background = element_blank(),
                       panel.border=element_rect(fill=NA),
                       panel.grid.major = element_blank(),
                       panel.grid.minor = element_blank(),
                       strip.background=element_blank(),
                       axis.title  = element_text(size = 20, colour = "black"),
                       axis.text.x=element_text(colour="black",size =20),
                       axis.text.y=element_text(colour="black",size =20),
                       axis.ticks=element_line(colour="black"),
                       plot.margin=unit(c(1,1,1,1),"line"),
                       plot.title = element_text(colour="black",size =20))


# Setup Variables 
volcano <- DEResults %>% .[-1,]
volcano$log_pval <- -log10(volcano$pvalue)
sig_up_data <- volcano %>% filter(log2FoldChange > 0) %>% filter(pvalue < 0.05)
sig_down_data <- volcano %>% filter(log2FoldChange < 0) %>% filter(pvalue < 0.05)

# Setup Variables for genes to label
UP_genes <- c("Mertk","Tgfbr2","Adam22","Nr3c1","Tgfb1","Tgfbr1","Spi1")
UP_genes_data <- sig_up_data %>% filter(gene %in% UP_genes)

DOWN_genes <- c("Clu","Ifngr2","Ctsz","Calm3","Apoe","Nfkb1","Lgals3","Ccr1","Ccl5","Clec7a","Ccl2")
DOWN_genes_data <- sig_down_data %>% filter(gene %in% DOWN_genes)



# Scatter Volcano Plot
scatter <- ggplot(volcano, aes(x= log2FoldChange, y= log_pval, label = gene)) +
  geom_point(size=0.5, color = 'grey90') +
  geom_point(data = sig_up_data, color = 'red', size = 1)+
  geom_point(data = sig_down_data, color = 'dodgerblue3', size = 1) +
  theme_scatter +
  xlim(-5,5)+
  ylim(0,10) +
  geom_vline(xintercept = 0.15,colour="grey", linetype = "longdash") +
  geom_vline(xintercept = -0.15,colour="grey", linetype = "longdash") +
  geom_hline(yintercept = 1.3, colour="grey", linetype = "longdash") +
  labs(title = table_header,
       x = "Log2FC",
       y = "-log(pvalue)") +
    geom_text_repel(data = UP_genes_data,
                  color = 'black', size = 6,fontface = 'italic',
                  min.segment.length = 0,
                  nudge_y = 1.5,
                  nudge_x = 2 - abs(UP_genes_data$log2FoldChange),
                  max.overlaps = Inf,
                  box.padding = unit(0.5, 'mm'),
                  point.padding = unit(0.5, 'mm'),
                  direction = "y") +
  geom_text_repel(data = DOWN_genes_data,
                   color = 'black',size = 6, fontface = 'italic',
                   min.segment.length =0,
                   nudge_y = 2,
                   nudge_x = -3 + abs(DOWN_genes_data$log2FoldChange),
                   max.overlaps = Inf,
                   point.padding = unit(0.5, 'mm'),
                  direction = "y")


# Save Image
ggsave(filename = paste0(workdir, expdir, "plots/VolcanoPlots/", fileprefix, "GOI.pdf"),
  plot = scatter,
  device = NULL,
  path = NULL,
  scale = 1,
  width = NA,
  height = NA,
  units = c("mm"),
  dpi = 300,
  limitsize = TRUE,
  bg = NULL
)
# View Volcano Plot
scatter 
```

## 2.4 Pathway Analysis
### 2.4.1 Perform EnrichGO

```{r}
##_______Important________##
## Use DESEQ results generated in Figure 1B section
## _______________________##

# Isolate out DEGs and pull gene names/log2foldchanges
DEG_genes <- sigDEResults %>% pull('gene')
DEG_Log2FC <-  sigDEResults %>% pull('log2FoldChange')

# Map gene names to ENTREZID 
DEG_genes <- mapIds(org.Mm.eg.db, DEG_genes, 'ENTREZID', 'SYMBOL')

# Assign names 
combined_DEG <-as.data.frame(cbind(DEG_genes, DEG_Log2FC))
names(DEG_Log2FC) = DEG_genes

# Run EnrichGO
go <- enrichGO(gene = DEG_genes,
               'org.Mm.eg.db',
               keyType = "ENTREZID",
               ont = "BP",
               pvalueCutoff = 0.05,
               pAdjustMethod = "BH")

# Set to Readable & filter
edox <- setReadable(go, 'org.Mm.eg.db', 'ENTREZID')
go_results <- edox@result %>% filter(pvalue < 0.05)


# Retrieve information on pathways of interest
go_Phago <- edox %>% filter(ID %in% c('GO:0006909')) %>% .@result %>% .$geneID %>% str_split(., pattern = "/") %>% .[[1]]
go_APP <- edox %>% filter(ID %in% c('GO:0019882')) %>% .@result %>% .$geneID %>% str_split(., pattern = "/") %>% .[[1]]
go_Interferon <- edox %>% filter(ID %in% c("GO:0034341"))  %>% .@result %>% .$geneID %>% str_split(., pattern = "/") %>% unlist() %>% unique()
go_neuroInf <- edox %>% filter(ID %in% c('GO:0150076')) %>% .@result %>% .$geneID %>% str_split(., pattern = "/") %>% .[[1]]
```

### 2.4.2 Generate Heatmaps (Supplementary Figure 3)

```{r}
# Filter sigDEcounts using GO results
Filtered_PhagoHeat <- sigDEcounts %>% filter(gene %in% go_Phago)
Filtered_Neuro <- sigDEcounts %>% filter(gene %in% go_neuroInf)
Filtered_APP <- sigDEcounts %>% filter(gene %in% go_APP)
Filtered_Interferon <- sigDEcounts %>% filter(gene %in% go_Interferon)


# Generate Heatmaps per BP 
genheatmap(Filtered_PhagoHeat, colnames(Filtered_PhagoHeat)[-1], height = 30, width = 10, directory = heatmapdir, cutrow = 2, gap_col = c(5),fileprefix = fileprefix, suffix = "_GO_Phagocytosis_Heatmap.pdf", show_row = TRUE)
genheatmap(Filtered_Neuro, colnames(Filtered_Neuro)[-1], height = 30, width = 10, directory = heatmapdir, cutrow = 2, gap_col = c(5),fileprefix = fileprefix, suffix = "_GO_NeuroInflammation_Heatmap.pdf", show_row = TRUE)
genheatmap(Filtered_APP, colnames(Filtered_APP)[-1], height = 30, width = 10, directory = heatmapdir, cutrow = 2, gap_col = c(5),fileprefix = fileprefix, suffix = "_GO_AntigenPres_Heatmap.pdf", show_row = TRUE)
genheatmap(Filtered_Interferon, colnames(Filtered_Interferon)[-1], height = 30, width = 10, directory = heatmapdir, cutrow = 2, gap_col = c(5),fileprefix = fileprefix, suffix = "_GO_ResponsetoIFNy_Heatmap.pdf", show_row = TRUE)
```







# 3.0 Figure 1C
## 3.1 Generate IPA DotPlot

```{r}
##________UPDATE ENTIRE SECTION___________##
# Update the Paths for IPA graph output directory
IPAgraphdir <- paste0(workdir)
fileprefix2 <- paste0(fileprefix,"_MC")
##________________________________________##

# Load IPA Data
IPAdata <- readxl::read_excel(path = "~/Dropbox (Partners HealthCare)/SK-5A4F_SN0268732_Wesley_Nov_2022_Xenon_Human_Mouse/results/Madison/APPPS1/SK-5A4F_APPPS1_Mouse_XenvsAir_pvalue01_pathways.xls",sheet = "Sheet0",skip = 1) %>%
  dplyr::mutate(Experiment = paste0("APP.PS1_XenVsAir")) %>% dplyr::rename(c("Ingenuity.Canonical.Pathways" = "Ingenuity Canonical Pathways"), "z.score" = "z-score")

# Setup Pathways of interest
path.of.interest <- c(
"Role of CHK Proteins in Cell Cycle Checkpoint Control",
"Senescence Pathway",
"Synaptogenesis Signaling Pathway",
"Autophagy",
"iNOS Signaling",
"IL-1 Signaling",
"CCR5 Signaling in Macrophages",
"Oxidative Phosphorylation")


# Generate Graph of Interest
graph <- IPApoigraph_single(data = IPAdata, POI = path.of.interest, 
                            directory = IPAgraphdir, fileprefix = fileprefix2, suffix = "poi")

# View Graph
graph
```


