# Code for Manuscript: 
This repository contains the analysis scripts and data used for the manuscript titled "[Manuscript Title]". Each script corresponds to a specific figure in the paper and is labeled accordingly. This code was used to generate all results and visualizations included in the manuscript.


## Files and Figures
The files are organized by figure number. Below is a brief description of the structure:

| Figure in Manuscript | Experiment Type | Script Name | GSE |
-----------------------|-----------------|------------|-------|
Fig.1 b-c | Xenon in APPPS1 - Mice | Fig1BC_APPPS1_Mouse_XenonvsAir | GSE269157 |
Fig.1 l-m | Xenon in Microglia in vitro - Mice |
Fig. 3 | Xenon scRNAseq Myeloid |
Fig 4. m-n | Xenon in IPSMG H5xFAD | 
Fig. 4 r, u and v | 5xFAD-MITRG Astrocytes 6M |
Fig. 6 a-g | Xenon in IFNgr1 cKO |
Fig. 6 i-k | Xenon in APPPS1 IFNy blocker-Mice |
Fig. 8  | Tau |
Supp. Fig.1 d-e | Xenon PD in MGnD paradigm |
Supp. Fig.2 a-f | Xenon PK in APP.PS1 microglia -mice |
Supp. Fig. 3 d-e | Tdt Tomato - mice |
Fig. 3 | Xenon scRNAseq Myeloid |
Supp Fig 5. a-d | Xenon-iMGL in vitro |
Supp Fig.6 a-c | 5xFAD-MITRG-iMGL 6M |
Supp Fig.7 c-e | Xenon in IFNy blocker -mice |
Supp Fig 9 a-g | Xenon PK in APP.PS1 Tcells -mice |

Each script contains all necessary steps to reproduce the analysis and figures presented in the manuscript. Be sure to follow the usage instructions and install the required dependencies.

## Languages and Packages
The code is primarily written in the following languages:
* R (v4.2.1): Used for data analysis and visualization. <br>
* Shell: Used for pre-processing the fastq files

### Major R Packages: <br>
* ggplot2 – for data visualization <br>
* Seurat – for single-cell RNA sequencing analysis <br>
* biomaRt – for querying gene information <br>
* fgsea – for gene set enrichment analysis <br>

## Contact
For any questions or further information, please contact:

* Madison Carpenter
* Email: mmcarpenter@bwh.harvard.edu
* GitHub: https://github.com/madison-car

## Citation 
